contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+  // +optional  
+  CapacityBytes *uint64 `json:"capacityBytes,omitempty"`  
+  // InodesFree represents the free inodes in the filesystem.  
+  // +optional  
+  InodesFree *uint64 `json:"inodesFree,omitempty"`  
+  // Inodes represents the total inodes in the filesystem.  
+  // +optional  
+  Inodes *uint64 `json:"inodes,omitempty"`  
+}  
+
+// PodUsage holds pod-level unprocessed sample stats.  
+type PodUsage struct {  
+  // UID of the pod  
+  PodUID string `json:"uid"`  
+  // Stats pertaining to pod total usage of cpu  
+  // This may include additional overhead not included in container usage statistics.  
+  // +optional  
+  CPU *CpuUsage `json:"cpu,omitempty"`  
+  // Stats pertaining to pod total usage of system memory  
+  // This may include additional overhead not included in container usage statistics.  
+  // +optional  
+  Memory *MemoryUsage `json:"memory,omitempty"`  
+  // Stats of containers in the pod.  
+  Containers []ContainerUsage `json:"containers" patchStrategy:"merge" patchMergeKey:"uid"`  
+  // Stats pertaining to volume usage of filesystem resources.  
+  // +optional  
+  Volumes []VolumeUsage `json:"volume,omitempty" patchStrategy:"merge" patchMergeKey:"name"`  
+}  
+
+// ContainerUsage holds container-level usage stats.  
+type ContainerUsage struct {  
+  // UID of the container  
+  ContainerUID string `json:"uid"`  
+  // Stats pertaining to container usage of cpu  
+  // +optional  
+  CPU *CpuUsage `json:"memory,omitempty"`  
+  // Stats pertaining to container usage of system memory  
+  // +optional  
+  Memory *MemoryUsage `json:"memory,omitempty"`  
+  // Stats pertaining to container rootfs usage of disk.  
+  // Rootfs.UsedBytes is the number of bytes used for the container write layer.  
+  // +optional  
+  Rootfs *DiskUsage `json:"rootfs,omitempty"`  
+  // Stats pertaining to container logs usage of Disk.  
+  // +optional  
+  Logs *DiskUsage `json:"logs,omitempty"`  
+}  
+
+// CpuUsage holds statistics about the amount of cpu time consumed  
+type CpuUsage struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // Total CPU usage (sum of all cores) averaged over the sample window.  
+  // The "core" unit can be interpreted as CPU core-nanoseconds per second.  
+  // +optional  
+  UsageNanoCores *uint64 `json:"usageNanoCores,omitempty"`  
+  // Cumulative CPU usage (sum of all cores) since object creation.  
+  // +optional  
+  UsageCoreNanoSeconds *uint64 `json:"usageCoreNanoSeconds,omitempty"`  
+}  
+
+// MemoryUsage holds statistics about the quantity of memory consumed  
+type MemoryUsage struct {  
+  // The time at which these stats were updated.  
+  Timestamp metav1.Time `json:"time"`  
+  // The amount of working set memory. This includes recently accessed memory,  
+  // dirty memory, and kernel memory.  
+  // +optional  
+  WorkingSetBytes *uint64 `json:"workingSetBytes,omitempty"`  
```

Comment:
Given that we are redoing the API, I'd recommend dropping `WorkingSet` prefix and instead stick with usage.